pkg/controller/infrastructure/infraflow/reconciler.go

```diff
@@ -0,0 +1,189 @@
+	reconciler, err := NewAzureReconciler(infra, cfg, cluster, f.factory)
+	if err != nil {
+		return nil, err
+	}
+	graph := f.buildReconcileGraph(reconciler)
+	fl := graph.Compile()
+	if err := fl.Run(ctx, flow.Opts{}); err != nil {
+		return nil, flow.Causes(err)
+	}
+	return reconciler.GetInfrastructureStatus(ctx)
+}
+
+func (f FlowReconciler) buildReconcileGraph(reconciler *azureReconciler) *flow.Graph {
+	whiteboard := shared.NewWhiteboard()
+
+	g := flow.NewGraph("Azure infrastructure reconcilation")
+	resourceGroup := f.addTask(g, "resource group creation", reconciler.ResourceGroup)
+
+	vnet := f.addTask(g, "vnet creation", reconciler.Vnet, shared.Dependencies(resourceGroup))
+
+	f.addTask(g, "availability set creation", reconciler.AvailabilitySet, shared.Dependencies(resourceGroup))
+
+	routeTable := f.addTask(g, "route table creation", func(ctx context.Context) error {
+		routeTable, err := reconciler.RouteTables(ctx)
+		whiteboard.Set(routeTableID, *routeTable.ID)
+		return err
+	}, shared.Dependencies(resourceGroup))
+
+	securityGroup := f.addTask(g, "security group creation", func(ctx context.Context) error {
+		securityGroup, err := reconciler.SecurityGroups(ctx)
+		whiteboard.Set(sGroupID, *securityGroup.ID)
+		return err
+	}, shared.Dependencies(resourceGroup))
+
+	ip := f.addTask(g, "ips creation", func(ctx context.Context) error {
+		ips, err := reconciler.PublicIPs(ctx)
+		if err != nil {
+			return err
+		}
+		err = reconciler.EnrichResponseWithUserManagedIPs(ctx, ips)
+		if err != nil {
+			return fmt.Errorf("enrichment with user managed IPs failed: %v", err)
+		}
+		whiteboard.SetObject(publicIPMap, ips)
+		return nil
+	}, shared.Dependencies(resourceGroup))
+
+	natGateway := f.addTask(g, "nat gateway creation", func(ctx context.Context) error {
+		ips := whiteboard.GetObject(publicIPMap).(map[string][]network.PublicIPAddress)
+		resp, err := reconciler.NatGateways(ctx, ips)
+		whiteboard.SetObject(natGatewayMap, resp)
+		return err
+	}, shared.Dependencies(ip))
+
+	f.addTask(g, "subnet creation", func(ctx context.Context) error {
+		routeTable := armnetwork.RouteTable{
+			ID: whiteboard.Get(routeTableID),
+		}
+		securityGroup := armnetwork.SecurityGroup{
+			ID: whiteboard.Get(sGroupID),
+		}
+		natGateway := whiteboard.GetObject(natGatewayMap).(map[string]armnetwork.NatGatewaysClientCreateOrUpdateResponse)
+		return reconciler.Subnets(ctx, securityGroup, routeTable, natGateway)
+	}, shared.Dependencies(securityGroup), shared.Dependencies(routeTable), shared.Dependencies(natGateway), shared.Dependencies(vnet))
+	return g
+
+}
+
+// TODO copy from AWS PR (use taskBuilder component to share?)
+// addTask adds a wrapped task for the given task function and options.
+func (f FlowReconciler) addTask(g *flow.Graph, name string, fn flow.TaskFn, options ...shared.TaskOption) flow.TaskIDer {
```

Comment:
Martin's flowContext has a persistor which I don't need and I cannot opt out. Since it's only very few code and since the libraries are not synced at the moment at any rate, I would defer the refactor to opt out of the persistor until the libraries are really shared